### ISSUE TITLE
Don't create pp3 files unless something has changed

### ISSUE THREAD

**Issue by neilpho** [Beep6581/RawTherapee#2857]
At the moment as I move through my files in the editor window a pp3 file is created for each one I look at. I think that such a file should only be created if I change something from the default profile (or I explicitly request the file to be created).

This would make the tick badge more useful by showing those files I have actually edited. It would stop unnecessary writes to disk. It would avoid writing the current defaults to a file, when they may change or improve in a future release.

**Comment by midluk** [comment]
I second this suggestion. It should of course be possible to save a pp3 without changing anything by explicitly saving the settings or by exporting the image, but by default it should not be saved. It happens quite often that I look at an image (for example to compare it to an image I'm currently adjusting the settings for) and afterwards have to clear the processing profile so I know I still have to properly edit the processing profile. At least as far as I see it, for this I have to click through two layers of menus.
Other option would be a simple "clear processing profile when changing away from image" button in the editor (next to the star rating and "Move to trash"). This might even be more useful than the original suggestion and does not change default behavior. This "clear processing profile" button should also be available in the File Browser, I think.

**Comment by Hombre57** [comment]
@midluk First, you're assuming that when switching to another image, it should loose the settings by default. We've chosen the other way round, by default, save the currently displayed settings (which might not be the last one in the History).

If you want to review your images based on the embedded full sized preview image (out of camera), you can now use the Inspector tab which has been made for that, on the right side of the File Browser panel.

Then, if you want to delete your settings of some images after they have been opened, then you have a contextual menu over the thumbnails for that (I don't remember which one exactly, but it should be something like "Reset xxx...").

@neilpho The sole fact of opening an image in the editor require to attribute some settings to see the RT version of the image, i.e. it switch the rendering of the thumbnail from "Embedded OOC preview" to "RT's version", with a corresponding pp3 file.

If you're still in review phase, I'd suggest to attribute at least one star or reject image by putting them in the trash, or attribute at least one star to the edited images. Yes, that's a workaround, but your demand is not that simple to do.

**Comment by midluk** [comment]
That ispector Tab might work on a huge 4K monitor where you have plenty of screen space, but at least on my 1680x1050 monitor it is unusable, because I see only a very small section of the image (can you actually zoom?) while at the same time the file browser looses its purpose of getting an overview over the images, because the thumbnail part has to be made incredibly small. And unfortunately the scroll state of the file browser does not seem to be synchronized with what image is currently selected in the editor, so you always have to search for the correct image.
I know the settings can be deleted by using three mouse clicks (right, left, left) and some careful mouse navigation in between, but that is definitely not user friendly and can not be properly done frequently. That is why I'm asking for a simple one-click button.

**Comment by neilpho** [comment]
I don't think you should be saving changes that a human hasn't made. Even if changes are made, if the result doesn't differ from the default it should not be saved.

The preview image in the RAW file is not full sized, so is not suitable for checking focus. It's slow to render a full image, but it is worth it to me, but you can see that I don't want extra work on top of it.

If I am reviewing 650 images from one day then I do not want 600 unnecessary  right click and remove pp3 operations.

I think it's just bad practice to write unnecessary stuff to disk. Also I cannot currently see which images I have actually worked on.

I think RT is a fantastic package. I think we should work on moving it closer to perfection.

**Comment by Hombre57** [comment]
The thing is that actually, we are very few to actively work on RT, and I currently work on a Spot Removal tool which you might be more interested in than fine tuning the workflow, where you can still find workarounds. I too wish it would be a perfect package :), but then more devs would have to chim in.

@midluk For reviewing my images, I find the Inspector very useful : I fold the left panel thanks to the "Fold left panel" button, then narrow the thumbnail part to have only one column of thumbs and a very large area for the Inspector. So yes, it's 100% of the embedded preview only, but it's very practucal because I only have to move the mouse over the thumb to have a magnified view. You can aslo adjust the number of image buffer (last x examined images ) for the Inspector tool in Preferences. On my side, my camera is set to save a full sized image in the DNG file, that I can batch extract to share with friends if they can't wait on the final development.

Anyway, your point is still valid, and I'll let this issue opened, if someone wants to take it.

**Comment by midluk** [comment]
I think I will try to fix this myself.
I have a working button to clear the processing profile, just have to figure out how to update the icon so it is grayed out (or invisible) when no processing profile is present to clear (at the moment it always reflects the state when the file browser entry was created).

**Comment by midluk** [comment]
Here is a first version of a patch implementing a clear profile button above the thumbnail.
The button is greyed out when no processing profile is present to be cleared. I'm using FileBrowserEntry::getIconsOnImageArea() to update the icon (not sure if it is OK to add this side effect to that function, but I did not find any other easy solution to achieve the result).
I will try to make the selection to clear the profile stick when the image is currently open in the editor, so no profile will be saved on changing to a different image. But this will definitely be more difficult than adding that simple button.

https://gist.github.com/midluk/37cbf16ea0c2585cb413
With two icons for light icon set
![clear-profile-grey](https://cloud.githubusercontent.com/assets/366536/9369562/56b14d76-46ca-11e5-9204-08cf31361932.png) clear-profile-grey.png
![clear-profile](https://cloud.githubusercontent.com/assets/366536/9369561/56b0b208-46ca-11e5-9331-5673107f8675.png) clear-profile.png
and two icons for the dark icon set
![clear-profile-grey](https://cloud.githubusercontent.com/assets/366536/9369583/785d98d0-46ca-11e5-804a-b70aca88e4e5.png) clear-profile-grey.png
![clear-profile](https://cloud.githubusercontent.com/assets/366536/9369584/78643230-46ca-11e5-9968-34d69ccc7dcf.png) clear-profile.png

**Comment by michaelezra** [comment]
FYI, thumbnail has a checkmark icon that indicates that image was edited (at least opened in the editor). There is also filtering in place to filter thumbnails based on this flag.

**Comment by Beep6581** [comment]
1- We should first decide on the logic:
1.1 - User opens a raw photo in the Editor, likes the result of the "default profile for raw photos" and wants to keep it, how does he close it?
1.2 - User opens a raw photo in the Editor, does not like the result of the "default profile for raw photos" and does not want to keep it, how does he close it?
1.3 - User opens a raw photo in the Editor, does not like the result of the "default profile for raw photos", plays with some sliders, then goes back in history to the original state deciding that he does in fact want to keep it.
1.5 - User batch-applies a profile to a selection of photos from the File Browser or Filmstrip, therefore the profiles are to be kept.
1.6 - Profiles are generated from the outside by a script and all are to be kept,
etc. Other possible situations involving METM and SETM differences.

2- Icons must be in the SVG format consistent in size and style with our other icons in tools/source_icons/

3- Patch style (indentation etc) must be consistent with existing style, so no tabs.

**Comment by Hombre57** [comment]
@midluk I don't think that the "-grey" icons are necessary. If unused, the icon should be grayed out by making it unsensitive, to stay consistent with other buttons in RT.

at all,
yes, the way that RT saves the pp3 file is not ideal, it does save the active parameter set when closing the tab or when switching to another image, + a shortcut to silently save the actual parameter set (and shortcut only).

To handle the different desire and workflow, we could add 3 buttons in the top toolbar of the Editor tab:
1. Profile generation strategy  (I would integrate that as a popup menu in #2 below in fact)
Opens a contextual menu with the following checkbox :
   o a. Generate when opening an image
   o b. Generate/Save current profile when closing tab or switching to another image

2. Save current parameter set
This would essentially do what Ctrl+S does, i.e. silently save the current parameter set. Once saved, the button would be made insensitive. That would tell the user if the image is saved or not.

3. Delete pp3
This button would delete the sidecar and cached profile when pressed. If option (b) above is checked, the pp3 file will be created again when closing the tab or switching to another image.

I think this is a good compromise and offers flexibility.

**Comment by Beep6581** [comment]
If such a thing was implemented, a profile should not be saved if deleting the profile is the last thing the user does before closing the Editor regardless of the strategy chosen.

I think Preferences is a better place for this choice of strategy, not as a button in any toolbar.

**Comment by midluk** [comment]
I based the icon on logoicon-wind.png which is used in the right click menu in the file browser for "processing profile operations" and does not seem to have an svg version. It is definitely possible to improve it, but I don't feel like I have the skills to do that myself.
Can you show me how I can instruct an LWButton to be grayed out? Would definitely be better than changing the icon, but I don't see that capability is currently implemented.

I fixed (or better: broke like in the rest of rt code) the indentation in that line I missed previously.

I think most important thing to implement is "a profile should not be saved if deleting the profile is the last thing the user does before closing the Editor regardless of the strategy chosen", which I don't think needs to wait for a decision on the rest of the logic. But I guess until I have found the part of the code where this can be done, the rest of the decisions will be have been done.

**Comment by Hombre57** [comment]
@midluk use the "set_sensitive(false)" method to grey out an icon.

If you don't mind, I'll take this issue, since it seems that it's more useful than I though.

**Comment by midluk** [comment]
@Hombre57 The LWButton in lwbutton.{h,cc} does not have a method set_sensitive(bool). Seems like you mix it up with Gtk::MenuItem
Feel free to take this issue.

**Comment by Hombre57** [comment]
I've updated your patch @midluk , but finally, I think that it's quite dangerous to have this icon here : if a user inadvertently click on it instead of the color label, there's no security, no confirmation and he will lost his editing that he spent so much time on it ! So we should either : 
1. add a confirmation box, but it'll be very annoying at some point
2. use the double-click event on this particular icon
3. keep only the menu entry to delete the profile, more secured because placed in a sub-menu
4. ...?

I'll do the Editor's new buttons anyway.

**Comment by Hombre57** [comment]
The pp3 is also saved when "Save as..." is clicked and an image has been generated (button on the bottom left of the preview image). Should we keep this behavior or be able to disable it through Preferences?

**Comment by Hombre57** [comment]
The pp3 is also saved when the image is sent to Queue. Should we be able to disable this ?

In the case of the "Save as" button, the output pp3 file can be generated as well, so the user still has a way to restore those parameters.

For the "send to queue" event, there's no way to get the parameters back, unless you effectively process the queue and chose to generate an output pp3 file.

**Comment by Hombre57** [comment]
The collateral of this new Save button, is that whenever you press the Save button and continue editing your image, the thumbnail will have to reflect the current state of the edition. Then you chose to close the Editor tab without saving (and auto save on close is disabled), then the thumbnail should revert to the last valid state, when the Save button has been clicked for the last time.

Is this what we want ? Of course, the patch will be more important and will take more time, but I don't think we should make things half done too. This patch will introduce a common save behavior, i.e. open the image, hit "Save" whenever you think your image is nice, and close eventually without saving, while still offering the RT's way.

**Comment by midluk** [comment]
Some backup of the processing parameters pp3.bak or pp3~ could be kept when clearing the processing profile, so they can be restored just in case (ideally with the button then offering this functionality). But these files should be kept track of and be deleted when RT is closed. Or just all matching backup files are deleted when leaving the directory in file browser or closing RT.

I also discovered not quite intuitive behavior when having an image ranked (star) and then clearing processing profile. When it is not ranked, it is restored to default profile when it is open in the editor (or no profile when it is not open in editor). When it is ranked, some other profile (I think neutral) is set.

**Comment by michaelezra** [comment]
My 2c. Overall i am satisfied that pp3 files are preserved and so far I never lost work within rawtherapee. To me this is the absolute highest priority. The absolute!;)

In my opinion, the easiest way to resolve confusion about pp3 files is to actually save them more frequently, not to choose when not to save them!:) Currently on 1st browse rt shows thums with embedded jpgs, not color managed. What should follow is a background low priority thread that applies default profile and assigns pp3s. Every image in RT should have a pp3. The Edited flag may need to tweaked so it would reflect that image was edited beyond the initial pp3 assignment -pp3 was pasted in or image opened in editor and something ( e.g. slider, or curve, or checkbox) was changed.

**Comment by Hombre57** [comment]
I don't think that it's worth to spend a lot of processing time to process images that are not yet reviewed, especially if you have hundreds of shots to review.

I've worked on this issue for some days now, because it's more complex that expected, i.e. I have to solve a misbehavior that is related to midluk's last comment, about not reverting to the embedded jpeg if the image has been tagged (rank, etc...). It's even worse: if you tag an image (left untouched), the embedded image is still displayed. Close RT and restart it -> the images that has been tagged are now processed.

That's why I'm refactoring the ParamsEdited class to make it more OO, and be able to handle everything properly.

Also, the Exif and IPTC section should be handled differently too I think : you can edit an image, and then affecting a PP3 to that image, then copy/partial paste the Exif/IPTC tags to other images. I think that if we only edit metatags, we shouldn't process the image, the tools has not been edited. Moreover, if we want to revert to the embedded image, do we necessarily revert the IPTC and Exif changes ? I'll certainly add a dedicated menu entry for that. If you're okay with that, what should be the behavior of the new Delete button on the thumbnails ? Delete only tools ? Or delete Tools + IPTC + EXIF ?

**Comment by Beep6581** [comment]
It seems to me that managing metadata changes would be easiest if Exif/IPTC/XMP changes were handled the same way as any tool changes, meaning we should be able to partial-paste them, and (the currently missing feature) to batch-set (or batch-clear) them from the File Browser.

If metadata changes will be handled separately from image tweaks, then the thumbnails will need a new overlay icon for metadata changes in addition to the green tick for image tweaks.

**Comment by Hombre57** [comment]
Do you mean partial-paste on a per-field granularity ?

Clearing the whole EXIF and IPTC is something that I'll add in this patch. The difficult thing is that they are embedded in the same PP3 file than the tool's parameters. Actually, when "Saving as" or "Loading" a pp3 (from Profile Panel), the rank, inTrash and color label are not save/loaded. But how should we handle the EXIF/IPTC changes ? Do we have to add 2 checkbox in the load/save dialog, or should we use the partial-load/save mechanism instead ? (which avoid me some coding ;) )

**Comment by michaelezra** [comment]
Saving exif and iptc in pp3 would have little value as it won't propagate to any other software one may use in the workflow. That's what XMP branch attempted - saving changes in standard XMP or writing into the image file itself. Tha last part was a problem as method of writing required overwriting the entire file which was an expensive IO operation. This ended up the primary difficulty of the XMP branch and that's where Fabio left it. Ideally, exif and iptc should be handled in industry-standard fashion as they are used for interoperability.

For partial paste, let's not go nuts and make it easier:); simply allow all exif / IPTC changes or none to be pasted. the same with reuse of the load mode button - would be better to simply reuse it and keep interface simpler.

**Comment by Beep6581** [comment]
> Do you mean partial-paste on a per-field granularity ?
No, just "Metadata", agreed with Michael.
A Metadata tab in the File Browser would be nice, but perhaps first we should address the open issues for the current metadata tab in the Editor to make it more usable.

For what its worth, I often use exiftool to add lens name, focal and aperture info into raws taken with manual non-electronic lenses.

**Comment by Hombre57** [comment]
AFAIK, XMP can be handled through a side car file, like the PP3 file, and it's rather standard, correct ? Since XMP can embed any kind of data, the RT's parameters could be saved in XMP files instead of PP3, the side car mechanism is already there (we should remove the cache though, since there is more cons than pros, and it applies to the PP3 files as well).

For now, and to avoid delivering too late and being out of synch with the Gtk3 branch, I'd suggest to continue with the PP3 file format.

@Beep6581 You use exiftool, but you know that you're kind of an exceptional man, do you  ? ;)

**Comment by Hombre57** [comment]
> For partial paste, let's not go nuts and make it easier:); simply allow all exif / IPTC changes or none to be pasted. the same with reuse of the load mode button - would be better to simply reuse it and keep interface simpler.

Sorry Michael, I don't understand what you mean.

**Comment by michaelezra** [comment]
Hi Hombre, I am simply for the simplest implementation. With partial paste dialog - to not provide the granular level of control for each exif/iptc field, but to allow user select holistically via only 2 check boxes 1. exif fields; 2 - iptc fields.

>  Do we have to add 2 checkbox in the load/save dialog, or should we use the partial-load/save mechanism instead ? (which avoid me some coding ;) )
To me the partial-load/save mechanism is sufficient:)

**Comment by Hombre57** [comment]
I'm almost done with the first patch. I still have pending question and I'll push it to a new branch.

- The Tool's params, the Exif params and the IPTC params can now be cleared separately.
- If the flags are to their default values, if the Tools, Exif and IPTC params are not set (or cleared), then the PP3 file is automatically deleted.
- If you paste or partial paste a profile that only contains Exif or IPTC parameters, no Tool's params will be created and the thumbnail will remain the one of the embedded image. The default "raw" params may contain Exif or IPTC default values, so they'll still be used for this partial paste.
- The green checkmark will be replaced by something that will show if Exif and IPTC are set (as suggested above), still to be done.

The "Delete params" button is still there (from midluk's last patch), but which key combination should we use to avoid pressing it mistakenly ? We can use any mouse button combination (including e.g. B1+B2) with modifier keys if you want. Also, pressing it will only delete the Tool's params. Should we have modifier keys to delete Exif and IPTC as well ? Keep in mind that the Ctrl and Shift modifier key + B1 are already used during thumb selection, so they're not really safe for this Delete button, unless we use B2 or B3.

I still have to check how it behaves if an image is edited in an Editor, and we click on this "Delete profile" button. I think it should delete the pp3 file as expected and bring back the embedded thumbnail, but when we close the Editor tab, the pp3 will logically be restored. This is a special "silly" situation that I don't want to spend too much time on.

**Comment by Beep6581** [comment]
Remembering which button combination to use for such a dangerous task would not be optimal, I would go for one button per delete category (delete tools/delete metadata/delete cache or whatever else) and yes, some safeguard for them, such as B1+B2 or ctrl+shift+click. Or a "Delete" combobox with tools/metadata/cache/whatever options, along with the safeguard.

What do you think about having the same safeguard work in combination with closing an editor tab or closing RawTherapee or switching to the next/previous image? For example, Ctrl+Shift+q would close RT without saving the PP3, Ctrl+Shift+w or Ctrl+Shift+click on the tab [x] would close the editor tab without saving the PP3, and Ctrl+Shift+next/previous buttons would do that without saving a PP3.

**Comment by Beep6581** [comment]
I tried the "SaveStrategy" branch (please avoid capitalization in the future) but it seems its not ready for testing yet - is that correct?

**Comment by Beep6581** [comment]
Is there any hope of this branch coming to fruition?

**Comment by Hombre57** [comment]
If you want a "yes" or "no", I'd say "maybe". I can have a look this w.e. (if I can find time) to update to the current dev (as a new branch). However I'll take time to finish the other issues I'm working on actually before finishing this one. Is this feature really necessary?

**Comment by Beep6581** [comment]
I've been using RT for a decade without it so it's not really necessary, but the idea sounds good, so I was just curious if it will still see the light of day. However if it complicates things too much then the cons may outweigh the pros.

**Comment by neilpho** [comment]
You can get used to a wooden leg, but life is better with a real one. I for one find the current behaviour unacceptable. This and cropping is why I don't feel comfortable recommending RT to people. 
It would be wonderful if you could fix it.

**Comment by Hombre57** [comment]
@Beep6581 @heckflosse @neilpho save-strategy2 branch created (I've deleted the previous one, too old and didn't managed to merge to current dev).

As explained in the commit, it crash when opening an image, I'll try to fix that asap.

**Comment by Floessie** [comment]
@Hombre57 I'd like to help you with the crash. Unfortunately, an icon is missing from your branch:

```
(gdb) catch throw 
Catchpoint 1 (throw)
(gdb) r
Starting program: /home/user/src/rawtherapee/build-debug/debug/rawtherapee 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe9dd8700 (LWP 5426)]
[New Thread 0x7fffe95d7700 (LWP 5427)]
[New Thread 0x7fffe8dd6700 (LWP 5428)]
[New Thread 0x7fffe3dfc700 (LWP 5429)]
[New Thread 0x7fffe35fb700 (LWP 5430)]
[New Thread 0x7fffe2199700 (LWP 5432)]
[New Thread 0x7fffe18fe700 (LWP 5433)]
[New Thread 0x7fffe0ced700 (LWP 5434)]

Thread 1 "rawtherapee" hit Catchpoint 1 (exception thrown), 0x00007ffff202826d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff202826d in __cxa_throw () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffff4788097 in Glib::FileError::throw_func(_GError*) () at /usr/lib/x86_64-linux-gnu/libglibmm-2.4.so.1
#2  0x00007ffff479bc6e in Glib::Error::throw_exception(_GError*) () at /usr/lib/x86_64-linux-gnu/libglibmm-2.4.so.1
#3  0x00007ffff3759b34 in Gdk::Pixbuf::create_from_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (filename=...)
    at pixbuf.cc:423
#4  0x0000555555e5d446 in RTImage::changeImage(Glib::ustring const&) (this=0x55555c0c7700, imageName=...) at /home/user/src/rawtherapee/rtgui/rtimage.cc:56
#5  0x0000555555e5d31e in RTImage::RTImage(Glib::ustring const&, Glib::ustring const&) (this=0x55555c0c7700, fileName=..., rtlFileName=..., __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /home/user/src/rawtherapee/rtgui/rtimage.cc:45
#6  0x0000555555be8864 in EditorPanel::EditorPanel(FilePanel*) (this=0x555559b706e0, filePanel=0x555557098400, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/user/src/rawtherapee/rtgui/editorpanel.cc:563
#7  0x0000555555e658f1 in RTWindow::createSetmEditor() (this=0x555556dc2600) at /home/user/src/rawtherapee/rtgui/rtwindow.cc:933
#8  0x0000555555e6177b in RTWindow::RTWindow() (this=0x555556dc2600, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/user/src/rawtherapee/rtgui/rtwindow.cc:214
#9  0x0000555555d2418e in (anonymous namespace)::create_rt_window() () at /home/user/src/rawtherapee/rtgui/main.cc:355
#10 0x0000555555d25cb9 in main(int, char**) (argc=1, argv=0x7fffffffe238) at /home/user/src/rawtherapee/rtgui/main.cc:667
(gdb) frame 4
#4  0x0000555555e5d446 in RTImage::changeImage (this=0x55555c0c7700, imageName=...) at /home/user/src/rawtherapee/rtgui/rtimage.cc:56
56	        const auto pixbuf = Gdk::Pixbuf::create_from_file (imagePath);
(gdb) print imageName 
$1 = (const Glib::ustring &) @0x7fffffffbb80: {static npos = 18446744073709551615, string_ = "clear-profile.png"}
```

Nevertheless, here's already a patch fixing some reorders:

```patch
diff --git a/rtgui/editorpanel.cc b/rtgui/editorpanel.cc
index 0f4cce42..70ab9828 100644
--- a/rtgui/editorpanel.cc
+++ b/rtgui/editorpanel.cc
@@ -483,14 +483,36 @@ public:
 
 };
 
-EditorPanel::EditorPanel (FilePanel* filePanel)
-    : catalogPane (nullptr), unmodified(true), realized (false), tbBeforeLock (nullptr), iHistoryShow (nullptr),
-      iHistoryHide (nullptr), iTopPanel_1_Show (nullptr), iTopPanel_1_Hide (nullptr), iRightPanel_1_Show (nullptr),
-      iRightPanel_1_Hide (nullptr), iBeforeLockON (nullptr), iBeforeLockOFF (nullptr), previewHandler (nullptr),
-      beforePreviewHandler (nullptr), beforeIarea (nullptr), beforeBox (nullptr), afterBox (nullptr),
-      beforeLabel (nullptr), afterLabel (nullptr), beforeHeaderBox (nullptr), afterHeaderBox (nullptr),
-      parent (nullptr), parentWindow (nullptr), openThm (nullptr), isrc (nullptr), ipc (nullptr),
-      beforeIpc (nullptr), err (0), isProcessing (false)
+EditorPanel::EditorPanel (FilePanel* filePanel) :
+    catalogPane(nullptr),
+    realized(false),
+    unmodified(true),
+    tbBeforeLock(nullptr),
+    iHistoryShow(nullptr),
+    iHistoryHide(nullptr),
+    iTopPanel_1_Show(nullptr),
+    iTopPanel_1_Hide(nullptr),
+    iRightPanel_1_Show(nullptr),
+    iRightPanel_1_Hide(nullptr),
+    iBeforeLockON(nullptr),
+    iBeforeLockOFF(nullptr),
+    previewHandler(nullptr),
+    beforePreviewHandler(nullptr),
+    beforeIarea(nullptr),
+    beforeBox(nullptr),
+    afterBox(nullptr),
+    beforeLabel(nullptr),
+    afterLabel(nullptr),
+    beforeHeaderBox(nullptr),
+    afterHeaderBox(nullptr),
+    parent(nullptr),
+    parentWindow(nullptr),
+    openThm(nullptr),
+    isrc(nullptr),
+    ipc(nullptr),
+    beforeIpc(nullptr),
+    err(0),
+    isProcessing(false)
 {
 
     epih = new EditorPanelIdleHelper;
diff --git a/rtgui/thumbnail.cc b/rtgui/thumbnail.cc
index 6a4885d4..ab59c778 100644
--- a/rtgui/thumbnail.cc
+++ b/rtgui/thumbnail.cc
@@ -37,10 +37,25 @@
 
 using namespace rtengine::procparams;
 
-Thumbnail::Thumbnail (CacheManager* cm, const Glib::ustring& fname, CacheImageData* cf)
-    : fname(fname), cfs(*cf), cachemgr(cm), ref(1), enqueueNumber(0), tpp(nullptr),
-      tagsSet(false), exifSet(false), iptcSet(false), paramsSet(false), defaultParamsSet(false), needsReProcessing(true),
-      imageLoading(false), lastImg(NULL), lastW(0), lastH(0), lastScale(0), initial_(false)
+Thumbnail::Thumbnail(CacheManager* cm, const Glib::ustring& fname, CacheImageData* cf) :
+    fname(fname),
+    cfs(*cf),
+    cachemgr(cm),
+    ref(1),
+    enqueueNumber(0),
+    tpp(nullptr),
+    tagsSet(false),
+    exifSet(false),
+    iptcSet(false),
+    paramsSet(false),
+    defaultParamsSet(false),
+    needsReProcessing(true),
+    imageLoading(false),
+    lastImg(nullptr),
+    lastW(0),
+    lastH(0),
+    lastScale(0),
+    initial_(false)
 {
 
     loadProcParams ();
@@ -60,10 +75,23 @@ Thumbnail::Thumbnail (CacheManager* cm, const Glib::ustring& fname, CacheImageDa
     tpp = nullptr;
 }
 
-Thumbnail::Thumbnail (CacheManager* cm, const Glib::ustring& fname, const std::string& md5)
-    : fname(fname), cachemgr(cm), ref(1), enqueueNumber(0), tpp(nullptr), tagsSet(false),
-      exifSet(false), iptcSet(false), paramsSet(false), defaultParamsSet(false), needsReProcessing(true),
-      lastW(0), lastH(0), imageLoading(false), lastImg(nullptr), initial_(true)
+Thumbnail::Thumbnail (CacheManager* cm, const Glib::ustring& fname, const std::string& md5) :
+    fname(fname),
+    cachemgr(cm),
+    ref(1),
+    enqueueNumber(0),
+    tpp(nullptr),
+    tagsSet(false),
+    exifSet(false),
+    iptcSet(false),
+    paramsSet(false),
+    defaultParamsSet(false),
+    needsReProcessing(true),
+    imageLoading(false),
+    lastImg(nullptr),
+    lastW(0),
+    lastH(0),
+    initial_(true)
 {
 
 
```

HTH,
Flössie

**Comment by Floessie** [comment]
@Hombre57 In the meantime, I copied another icon to that missing location. All works great, I get no crash at all when loading a DNG using the debug built in GDB. Am I missing something?

**Comment by Hombre57** [comment]
@Floessie 

> Am I missing something?

No, you just solved the bug (will check that when back home). It's possible that the icon were lost in the battle of manual merging. Unfortunately the crash was so severe here (Win7/MSYS2) that there was no bt available. Thanks for having fixed it, I'll try to find that icon or recreate it.

In SaveStrategy branch (first try), I even spread the `set`, `combine` and `initFrom` methods into each Params, to make all tools more OO. I didn't wanted to break too much things in other branch so didn't made that again.

**Comment by Hombre57** [comment]
@Floessie Your patch is committed.

**Comment by Hombre57** [comment]
Dev merged into `save-strategy2`, which got a new commit prior to this. It's on the right track, but still has minor things to solve.

@neilpho @Beep6581 @midluk @heckflosse @Floessie : I'm planning to update the thumbnail only when the user manually save its work thanks to the new Save button (or the existing shortcut). Anyone against that ? I don't think it makes sense to let the thumbnail represent a processing that is not saved yet.

**Comment by gaaned92** [comment]
@Hombre57 
W10 : on starting it changes some parameters of the option file: theme, disposition, default profile...

**Comment by Hombre57** [comment]
@gaaned92 Commit 822a2c9 should fix the problem.

**Comment by gaaned92** [comment]
@Hombre57  yes, fixed.

**Comment by Hombre57** [comment]
The save-strategy2 branch is almost finished in a feature point of view. I'll remove the debug outputs before merging.

How does it works now (use this for documentation) :

-----------------------------------

There is two new options in Preferences now (one was already ther in the options file, w/o GUI) :
   1. "**Save profile when closing the Editor or when switching to another image**" : if checked, the profile will be saved before closing the tab (e.g. when closing RT) or before loading a new image. In this case, thumbnails will show instantly each new adjustment, otherwise it will only updates when the Save button is pressed. Warning: if RT crash, the work will be lost if you didn't saved it manually or closed the tab, even though the thumbnail were showing the adjustments (same warning for actual dev).
   2. "**Save profile even if leaved untouched in the Editor**" : as it suggest, if no action is done in the Editor, the profile is not saved. This mean that if a file is being edited for the first time, the generated profile will not be stored. If the procparams is cleared from the PP3 file through the thumbnails' Delete button or the Editor's Delete button, the thumbnail will restore the embedded image and the Editor will build the procparam again through default profile, Custom Profile Builder and Dynamic Profile mechanisms (not fully tested). Selecting an entry in the History, would it be the "Profile cleaned" entry, will be considered as modifying the image and will be stored if checkbox of point 1 is checked.

To delete the associated params without clearing the tags and ratings, you have to press on Ctrl + Middle mouse button on the Delete buttons (in thumbnails or in Editor's toolbar), for security reasons.

-----------------------------------

I see only one remaining thing to add, but would like to have your comment on that : if first checkbox is unchecked, you'll lose all work done after having pressed the Save button when closing a tab or loading a new image. Usually, an alert box warns the user that work will be lost, and let him save it before exiting. I didn't made that warning alert, is it necessary in our context ? (I think so).

By default, the new behavior is the same as the old one, i.e. save automatically when closing tab or loading new image.

**Comment by Beep6581** [comment]
Right-clicking on a thumb, none of the Processing Profile Operations > Clear * work.
Also, RT quits with a `Segmentation fault (core dumped)` if I open a file and then close RT.

**Comment by Beep6581** [comment]
So, the segfault. Open `amsterdam.pef`, close RT, crash. It uses the neutral PP3. Single Editor Tab Mode - Vertical Tabs.
```
Thread 1 "rawtherapee" received signal SIGSEGV, Segmentation fault.
0x00007ffff70d5915 in gtk_widget_set_sensitive () from /usr/lib64/libgtk-3.so.0
(gdb) bt full
#0  0x00007ffff70d5915 in gtk_widget_set_sensitive () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#1  0x0000555555c2fa12 in EditorPanel::close (this=0x555559ab1000) at /home/morgan/programs/code-rawtherapee/rtgui/editorpanel.cc:1126
        is = 0x1f410437bb22700
#2  0x0000555555c2e756 in EditorPanel::~EditorPanel (this=0x555559ab1000, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /home/morgan/programs/code-rawtherapee/rtgui/editorpanel.cc:914
No locals.
#3  0x0000555555c2ec5e in EditorPanel::~EditorPanel (this=0x555559ab1000, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /home/morgan/programs/code-rawtherapee/rtgui/editorpanel.cc:964
No locals.
#4  0x00007ffff5353f43 in g_datalist_clear () from /usr/lib64/libglib-2.0.so.0
No symbol table info available.
#5  0x00007ffff566c1da in g_object_unref () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#6  0x00007ffff6fa995e in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#7  0x00007ffff6ebde5e in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#8  0x00007ffff56658a9 in g_closure_invoke () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#9  0x00007ffff567d73e in ?? () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#10 0x00007ffff568645c in g_signal_emit_valist () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#11 0x00007ffff568680b in g_signal_emit () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#12 0x00007ffff70cf23a in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#13 0x00007ffff566dfb0 in g_object_run_dispose () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#14 0x00007ffff70d8b09 in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#15 0x00007ffff3dff478 in Gtk::Container_Class::forall_vfunc_callback(_GtkContainer*, int, void (*)(_GtkWidget*, void*), void*) () from /usr/lib64/libgtkmm-3.0.so.1
No symbol table info available.
#16 0x00007ffff6ebde5e in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#17 0x00007ffff566597d in g_closure_invoke () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#18 0x00007ffff567d73e in ?? () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#19 0x00007ffff568645c in g_signal_emit_valist () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#20 0x00007ffff568680b in g_signal_emit () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#21 0x00007ffff70cf23a in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#22 0x00007ffff70dfcd3 in ?? () from /usr/lib64/libgtk-3.so.0
No symbol table info available.
#23 0x00007ffff566dfb0 in g_object_run_dispose () from /usr/lib64/libgobject-2.0.so.0
No symbol table info available.
#24 0x00007ffff3f109f8 in Gtk::Window::_release_c_instance() () from /usr/lib64/libgtkmm-3.0.so.1
No symbol table info available.
#25 0x00007ffff3f114cb in Gtk::Window::~Window() () from /usr/lib64/libgtkmm-3.0.so.1
No symbol table info available.
#26 0x0000555555ef89e0 in RTWindow::~RTWindow (this=0x5555571c9c30, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /home/morgan/programs/code-rawtherapee/rtgui/rtwindow.cc:285
No locals.
#27 0x0000555555ef8aac in RTWindow::~RTWindow (this=0x5555571c9c30, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /home/morgan/programs/code-rawtherapee/rtgui/rtwindow.cc:299
No locals.
#28 0x0000555555dbe3f8 in std::default_delete<RTWindow>::operator() (this=0x7fffffffcf90, __ptr=0x5555571c9c30) at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/unique_ptr.h:76
No locals.
#29 0x0000555555dbde7f in std::unique_ptr<RTWindow, std::default_delete<RTWindow> >::~unique_ptr (this=0x7fffffffcf90, __in_chrg=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/unique_ptr.h:236
        __ptr = @0x7fffffffcf90: 0x5555571c9c30
#30 0x0000555555dbcc54 in main (argc=1, argv=0x7fffffffd5c8) at /home/morgan/programs/code-rawtherapee/rtgui/main.cc:667
        m = <incomplete type>
        rtWindow = std::unique_ptr<RTWindow> containing 0x5555571c9c30
        exname = "/home/morgan/programs/code-rawtherapee/build/debug/rawtherapee", '\000' <repeats 449 times>
        exePath = {static npos = 18446744073709551615, string_ = "/home/morgan/programs/code-rawtherapee/build/debug"}
        ret = 0
```

**Comment by SaturnusDJ** [comment]
Kick.

How's the thought about the original poster's ( @neilpho ) idea now?
It's indeed annoying that for just opening images a .pp3 is created already. Now I don't know anymore which files I really worked on and which ones I just opened.

**Comment by Thanatomanic** [comment]
Can I chime in for a bit here? I've read through the entire conversation, and I fully agree with the original suggestion: only create a profile when an image gets edited. Additionally, I think auto-save is a wonderful feature to make sure you never lose your work. But after reading @Hombre57's last big post, I feel the current proposed strategy is making things a little more complicated than necessary.

> "Save profile when closing the Editor or when switching to another image" : if checked, the profile will be saved before closing the tab (e.g. when closing RT) or before loading a new image. In this case, thumbnails will show instantly each new adjustment, otherwise it will only updates when the Save button is pressed. Warning: if RT crash, the work will be lost if you didn't saved it manually or closed the tab, even though the thumbnail were showing the adjustments (same warning for actual dev).

So, if I would uncheck this box, no profile is ever saved, unless I explicitly Save it? I don't see a use case for this to be honest, but please enlighten me :smile: I mean, as soon as you _have_ edited a picture, auto-saving of the pp3 to not lose your work seems logical to me. And whenever you don't like your edit, there's already three options to start over using the context menu of a thumbnail: Processing profile operations > apply the Default profile, or Reset to default or Clear. And maybe in the future, when the entire editing history is preserved, you can simply use that to get back to an earlier editing point.
Also, but maybe I overlooked, this option was not really requested in this issue. So I would keep it simply and stay with the current behavior and do not introduce a dedicated Save profile button.

> "Save profile even if leaved untouched in the Editor" : as it suggest, if no action is done in the Editor, the profile is not saved. (...)

Fully agree, because that seems to be the essence: don't make a file when it's not needed.

> I see only one remaining thing to add, but would like to have your comment on that : if first checkbox is unchecked, you'll lose all work done after having pressed the Save button when closing a tab or loading a new image. Usually, an alert box warns the user that work will be lost, and let him save it before exiting. I didn't made that warning alert, is it necessary in our context ? (I think so).

If checkbox 1 is implemented, an alert box would certainly be user friendly. But please have an option to disable further notifications of this particular box if the user knows what he's doing and doesn't want to click the box every time he moves away from an image.

> By default, the new behavior is the same as the old one, i.e. save automatically when closing tab or loading new image.

Having the second checkbox unticked was the entire point of this issue, right? Imo, to have that by default would not break backwards compatibility in any way. Who is dependant on the pp3's being made for unedited images?

**Comment by doak** [comment]
Just want to add my two cents regarding an (as far as I see yet not mentioned) use case for that feature. I am quite new to *RawTherapee* (and even processing raw images at all), thus please forgive me if I overlooked some obvious workflow.

To process a bunch of similar pictures (with similar lightning conditions), I prepare a `base.pp3` which I want to apply to these pictures. Before running this batch job, I adjust selected images manually like changing exposure or rotating it *and* save these selected changes (i.e. without any other) stuff in the sidecar file next to the raw picture.
Afterwards I develop the JPEGs with e.g. `rawtherapee-cli -s -Y -p base.pp3 -c *.raf`. The switch `-s` ensures that the manually chosen parameters are merged into `base.pp3` (or the other way round, I don't know).  If I want to adjust other parameters or other pictures, I want to just update the sidecar files.

But since *RawTherappe* just creates and even overwrites any existing sidecar file if the corresponding raw file is opened, these manual adjustments are lost -- rendering that workflow useless.

Is this in your opinion a valid use case or is my workflow broken?

> "Save profile when closing the Editor or when switching to another image" 

As far as I understand, the mentioned option would enable the described use case, wouldn't it?

Furthermore, I quickly tested to enable the (already existing) option 'Save processing file to the cache' (mentioned in #4685), which seems to work as well. Although imho this is more like a workaround(, though a good one).

**Comment by Hombre57** [comment]
@doak 

> these manual adjustments are lost

Which manual edjustment are you referring to ? Opening a new file is manually editing it. I may have missed something in your explanation.

**Comment by doak** [comment]
@Hombre57, I mean the adjustment I did before like rotating the picture.

Let's assume I want to have the image rotated additionally to the "default" parameters I set in `base.pp3` for these bunch of pictures. My workflow is to configure this rotating and save (only this parameter) to the sidecar file. I use the ctrl+click feature of the 'save current profile' button to accomplish that. The resulting `<picture-name>.raf.pp3` file has only a few bytes.

With the above mentioned command (`rawtherapee-cli -s -Y -p base.pp3 -c *.raf`) I would be able to process the bunch of files with `base.pp3`, whereby existing sidecar files are merged into this profile.
Unfortunately the "manual adjustment" (i.e. the rotation in the example) will be overwritten (with the whole current profile) as soon as I close the picture. Even if I would work around that, as soon as I open and close one of the pictures again (without any modification), sidecar files will be created and existing ones (containing the "manual adjustments") will be replaced by full blown profiles (11kB) with all but the rotation disabled.

The issue now is:
  * Merging of the profile (during the mentioned command) is pointless if the sidecar is a full blown profile. There is no way to tell if something was disabled intentionally, or it is just the default setting.
  * You can't tell which picture was adjusted manually, because every one will get a sidecar file unconditionally.
  * You can't have a look which adjustments you did. This is quite simple if the sidecar files would only contain the selected parameters (like it did when saving it with ctrl-click like described before).
  * It would be even very easy to edit these "stripped-down" sidecar files by hand to throw away adjustments.

Possible improvements:
  * Sidecar files are only saved explicitly (like e.g. a word document, see below)
  * Implicitly saved sidecar files contain only touched parameters.

As far as I overlook this issue, the above mentioned solutions have been already discussed in this issue.
I would prefer the first one, because it is easier to implement (probably no or only minor adaptions to internal data model are needed) and contains less inherent magic which could (and probably will) confuse users. Furthermore it is consistent to usual tools like word processing: a text file is not saved by default, but will warn if the editor is closed and there are unsaved modifications. This logic could be easily extended to "don't warn if parameters have been saved manually (with ctrl+click) and profile has not been changed afterwards".

Be warned that I don't know *RawTherapee* very well and had no look into the sources. Furthermore I am not sure if perhaps my desired workflow is (let's say) strange.

Hopefully my explanation is clearer now.

**Comment by Hombre57** [comment]
@doak Thanks, it's very clear now.

Unfortunately, the requested modifications are not so simple in RT, because once you tag the file with stars and/or color label, and even with iptc and exif changes, it create a full blown profile. So the new mechanism would have created a profile anyway for stars and color label, but still flagged the image as unedited. The proposed patch here was almost done, i.e. seemed correct but did not passed the testing phase :disappointed: 

I understand that this is annoying (and it is for me too), but I can still live with that. Not that it's not important for you, but we're concentrating dev's manpower for more awaited features like local editing.

However there's an easy solution in your case : just save in your `base.pp3` all parameters but the geometry one (i.e. don't save `Rotation`, `Perspective`, `Distortion`, `Coarse transform`, etc...).

With the same command line, the full blown sidecar file will be read with neutral adjustment* + rotating. Then the `base.pp3` side car will replace the non geometrical parameters.

The more graphical way of doing what you want, in the File Browser panel, is to copy/paste from your reference image to the other one of the series (you can even _partial-copy_ and _partial-paste_ btw), then adjust the coarse rotation, and either put everything in the queue or use `rawtherapee-cli`.

I know it doesn't fully answer to your request (like being able to edit your file manually and see what is changed), but I'm sorry to say that this issue is low on our priority list.

\* _if you've selected `neutral` as default profile for raw image editing_

**Comment by doak** [comment]
@Hombre57, thanks you very much for you answer.

> [...] because once you tag the file with stars and/or color label [...]

I see. That means in fact major changes in data model would be needed.

> Not that it's not important for you, but we're concentrating dev's manpower for more awaited features like local editing.

That's absolutely fine.
I guess pull requests or rework of current patches are welcomed :)  (I am sorry, can't volunteer right now ...)

Thanks, too, for the details about the workaround.
